glibc: allow environment variable for /etc/ files (i.e. /etc/protocols) (#137601)
NixOS/nixpkgs: doc/builders/packages/etc-files.section.md

# /etc files {#etc}

Certain calls in glibc require access to runtime files found in /etc such as `/etc/protocols` or `/etc/services` -- [getprotobyname](https://linux.die.net/man/3/getprotobyname) is one such function.

On non-NixOS distributions these files are typically provided by packages (i.e. [netbase](https://packages.debian.org/sid/netbase)) if not already pre-installed in your distribution. This can cause non-reproducibility for code if they rely on these files being present.

If [iana-etc](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.iana-etc.x86_64-linux) is part of your _buildInputs_ then it will set the environment varaibles `NIX_ETC_PROTOCOLS` and `NIX_ETC_SERVICES` to the corresponding files in the package through a _setup-hook_.

```bash
❯ nix-shell -p iana-etc

[nix-shell:~]$ env | grep NIX_ETC
NIX_ETC_SERVICES=/nix/store/aj866hr8fad8flnggwdhrldm0g799ccz-iana-etc-20210225/etc/services
NIX_ETC_PROTOCOLS=/nix/store/aj866hr8fad8flnggwdhrldm0g799ccz-iana-etc-20210225/etc/protocols
```

> Nixpkg's version of [glibc](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.glibc.x86_64-linux) has been patched to check for the existence of these environment variables. If the environment variable are *not set*, then it will attempt to find the files at the default location within _/etc_.

Reviewer: Why do we link to hydra? Hydra does not contain any useful information except the build status.

Reviewer 2: How do we link to packages instead?

Reviewer: I would link to the file in git.

Reviewer 2: You mean a link to github?

Reviewer: yeah, I would do that.

Reviewer 2: https://github.com/NixOS/nixpkgs/pull/138984

> # /etc files {#etc}

Author: @Mic92 trying to be a "good" contributor and add docs with it.

Certain calls in glibc require access to runtime files found in /etc such as `/etc/protocols` or `/etc/services`

> [getprotobyname](https://linux.die.net/man/3/getprotobyname) is one such function

On non-NixOS distributions these files are typically provided by packages (i.e. _netbase_) if not already pre-installed in your distribution.
This can cause some non-reproducibility for some code if they rely on these files being present.

> The simplest way to add these files to your dependency closure is by exporting the `NIX_ETC_PROTOCOLS` or `NIX_ETC_SERVICES` environment variable pointing it to the [iana-etc](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.iana-etc.x86_64-linux) package.

Reviewer 2: ```suggestion
The simplest way to add [iana-etc](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.iana-etc.x86_64-linux) to your buildInputs:

<example-here>

This will set `NIX_ETC_PROTOCOLS` and `NIX_ETC_SERVICES` environment variables.
```

Author: Do you want me to write an example with `buildInputs` And explicitly mention the hook?

```nix
mkShell {
    name = "exmaple";

>     NIX_ETC_PROTOCOLS = "${pkgs.iana-etc}/etc/protocols";

Reviewer 2: We should definitely document this variable but at the same time I would also add a `setup-hook` to iana-etc to automatically set `NIX_ETC_PROTOCOLS` and `NIX_ETC_SERVICES` if it is added to the buildInputs.
As iana-etc also causes some rebuilds we should just add this setup-hook here right away.

Author: done!
